src/template/diagnostic.rs

```diff
@@ -0,0 +1,424 @@
+//! The code for error/diagnostics handling is heavily inspiered by
+//! [rust's](https://github.com/rust-lang/rust) compiler. While some code is adpated for use with
+//! punktf, some of it is also a plain copy of it.
+//!
+//! Specifically from those files:
+//! - https://github.com/rust-lang/rust/blob/master/compiler/rustc_span/src/lib.rs
+//! - https://github.com/rust-lang/rust/blob/master/compiler/rustc_span/src/analyze_source_file.rs
+//! - https://github.com/rust-lang/rust/blob/master/compiler/rustc_parse/src/parser/diagnostics.rs
+//! - https://github.com/rust-lang/rust/blob/master/compiler/rustc_errors/src/diagnostic.rs
+//! - https://github.com/rust-lang/rust/blob/master/compiler/rustc_errors/src/diagnostic_builder.rs
+//! - https://github.com/rust-lang/rust/blob/master/compiler/rustc_errors/src/emitter.rs
+
+use std::borrow::Cow;
+use std::collections::btree_map::Entry;
+use std::collections::{BTreeMap, HashSet};
+
+use color_eyre::owo_colors::OwoColorize;
+
+use super::source::Location;
+use super::span::ByteSpan;
+use crate::template::source::Source;
+
+#[derive(Default, Debug, Clone, PartialEq, Eq)]
+pub struct DiagnosticSpan {
+	/// A primary span is displayed with `^` bellow the spanned text.
+	pub(super) primary: Vec<ByteSpan>,
+
+	/// A label is displayed as the spanned text together with the label.
+	pub(super) labels: Vec<(ByteSpan, Cow<'static, str>)>,
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub enum DiagnositicLevel {
+	Error,
+	Warning,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct Diagnositic {
+	level: DiagnositicLevel,
+	msg: Cow<'static, str>,
+	span: Option<DiagnosticSpan>,
+	description: Option<Cow<'static, str>>,
+}
+
+impl Diagnositic {
+	pub fn new<M: Into<Cow<'static, str>>, D: Into<Option<impl Into<Cow<'static, str>>>>>(
+		level: DiagnositicLevel,
+		msg: M,
+		span: Option<DiagnosticSpan>,
+		description: D,
+	) -> Self {
+		Self {
+			level,
+			msg: msg.into(),
+			span,
+			description: description.into().map(|d| d.into()),
+		}
+	}
+
+	pub fn emit(&self, source: &'_ Source<'_>) {
+		let mut fmt = DiagnositicFormatter::new(source, &self.msg);
+
+		if let Some(span) = &self.span {
+			for primary in &span.primary {
+				fmt.primary_span(primary);
+			}
+
+			for (span, label) in &span.labels {
+				fmt.label_span(span, label);
+			}
+		}
+
+		if let Some(description) = &self.description {
+			for line in description.lines() {
+				fmt.description(line);
+			}
+		}
+
+		let out = fmt.finish();
+
+		match self.level {
+			DiagnositicLevel::Error => {
+				log::error!("{}{} {}", "error".bright_red().bold(), ':'.bold(), out)
+			}
+			DiagnositicLevel::Warning => log::warn!("{}", out),
+		};
+	}
+
+	pub fn level(&self) -> &DiagnositicLevel {
+		&self.level
+	}
+}
+
+#[must_use]
```

Comment:
Why would we need that? Isn't every struct member checked by default, whether it's used?

Reply:
Yeah i am battling with this one. My intention and as far as i understand it,
is that it should throw a warning if a `DiagnosticBuilder` is constructed but
not consumed (meaning `.build()` was not called). This is just meant for the
developer as a hint that they are likely using the API in a unintended way. But
so far i have not gotten it to work.

Reply:
lets remove it